### ISSUE TITLE
PLEASE add cdist

### ISSUE THREAD

**Issue by bionicles** [tensorflow/tensorflow#30659]
<em>Please make sure that this is a feature request. As per our [GitHub Policy](https://github.com/tensorflow/tensorflow/blob/master/ISSUES.md), we only address code/doc bugs, performance issues, feature requests and build/installation issues on GitHub. tag:feature_template</em>

**System information**
- TensorFlow version (you are using): 2.0b1
- Are you willing to contribute it (Yes/No): Yes

**Describe the feature and the current behavior/state.**
calculating distance matrices efficiently with tensorflow is a huge pain involving reading tons of stack overflow threads and re-implementing the same stuff. the solutions on stack overflow only cover euclidean distances and give MxM matrices even if you want city-block distance and MxMxD tensors ... it is extremely frustrating to experiment with optimal transport theory with tensorflow when such an obvious thing as CDIST is missing. 

**Will this change the current api? How?**
Yes, it will be easy to calculate distance matrices

**Who will benefit with this feature?**
anyone doing molecular dynamics, optimal transport, wasserstein, shape matching, graph theory, etc etc

**Any Other info.**
PDIST is also potentially useful
https://docs.scipy.org/doc/scipy/reference/generated/scipy.spatial.distance.cdist.html
https://docs.scipy.org/doc/scipy/reference/generated/scipy.spatial.distance.pdist.html

Block-wise version would be IDEAL as huge distance matrices often overflow gpu memory

**Comment by bionicles** [comment]
best i can do right now ... but it's impossible to do big problems ... im looking into tensor decomposition / factorization / compression to solve memory issues. ugh
```
@tf.function(input_signature=[tf.TensorSpec(shape=(None, D_FEATURES), dtype=tf.float16),
                              tf.TensorSpec(shape=(None, D_FEATURES), dtype=tf.float16)])
def get_distances(a, b):
    return tf.math.abs(tf.expand_dims(a, axis=1) - tf.expand_dims(b, axis=0))
```

**Comment by bionicles** [comment]
https://github.com/ftonolini45/Variational_Sparse_Coding/blob/master/models/vae_utils.py

```
def cartesian(arrays, out=None):
    """
    Generate a cartesian product of input arrays.
    FROM http://stackoverflow.com/questions/1208118/using-numpy-to-build-an-array-of-all-combinations-of-two-arrays/1235363#1235363
    Parameters
    ----------
    arrays : list of array-like
        1-D arrays to form the cartesian product of.
    out : ndarray
        Array to place the cartesian product in.
    Returns
    -------
    out : ndarray
        2-D array of shape (M, len(arrays)) containing cartesian products
        formed of input arrays.
    Examples
    --------
    >>> cartesian(([1, 2, 3], [4, 5], [6, 7]))
    array([[1, 4, 6],
           [1, 4, 7],
           [1, 5, 6],
           [1, 5, 7],
           [2, 4, 6],
           [2, 4, 7],
           [2, 5, 6],
           [2, 5, 7],
           [3, 4, 6],
           [3, 4, 7],
           [3, 5, 6],
           [3, 5, 7]])
    """

    arrays = [np.asarray(x) for x in arrays]
    dtype = arrays[0].dtype

    n = np.prod([x.size for x in arrays])
    if out is None:
        out = np.zeros([n, len(arrays)], dtype=dtype)

    m = n / arrays[0].size
    out[:,0] = np.repeat(arrays[0], m)
    if arrays[1:]:
        cartesian(arrays[1:], out=out[0:m,1:])
        for j in xrange(1, arrays[0].size):
            out[j*m:(j+1)*m,1:] = out[0:m,1:]
    return out
```

**Comment by bionicles** [comment]
```
def get_distances(xyz):
    xyz = tf.squeeze(xyz, 0)
    d = tf.reshape(tf.reduce_sum(xyz * xyz, 1), [-1, 1])
    return d - 2 * tf.matmul(xyz, xyz, transpose_b=True) + tf.transpose(d)
```

**Comment by bionicles** [comment]
docs for desired method have link to source: 
https://docs.scipy.org/doc/scipy/reference/generated/scipy.spatial.distance.cdist.html

**Comment by bionicles** [comment]
these both work for us now, might require some tf.squeeze action for other applications 

```
# @tf.function
# def get_distances(a, b):  # L1
#     print("tracing get_distances")
#     return B.sum(B.abs(tf.expand_dims(a, 0) - tf.expand_dims(b, 1)), axis=-1))

@tf.function
def get_distances(a, b):  # L2
    print("tracing get_distances")
    return B.sqrt(B.sum(B.square(tf.expand_dims(a, 0) - tf.expand_dims(b, 1)), axis=-1))
```
highly cool if fast distances were supported by TF team, also Sinkhorn / OT and Higher Order SVD, Normalized Compression Distance could be good loss for many many applications, but there's not a clear way to compress tensors for this metric in TF 2.0

**Comment by alextp** [comment]
Maybe this belongs in tf probability of tf addons, but unclear if it belongs in core TF given how easy it is to implement with TF.

**Comment by orybkin** [comment]
@alextp note that this is not easy to implement at all. The version provided by bionicles uses much more memory than necessary to compute cdist as for pairwise distances between two sets of N and M objects in needs an array NxM. This is quadratic memory scaling, and it blows up very easily.

Imagine you had 100 images 256x256 pixels each, and you had a batch of 20 such sequences, which is fairly reasonable. If you wanted to compute the CDIST of the images in the sequence, the forward + backward pass of this operation would not fit on your GPU. However, it is in fact only necessary to store 100 x 256 x 256 x 20 values, which is well below a gigabyte. This is what scipy implements, and it far from easy for an average user.

**Comment by alextp** [comment]
I believe cdist can be implemented as a composition of existing TF operations (efficient cdist amounts to a batch matmul under the hood, right?), so it can live in addons.

@rmlarsen do you know if I'm being confused here?

**Comment by orybkin** [comment]
I think that's right, for L2 distances a batchmul will do (but for other common distances it can be more tricky). I just wanted to note that it is rather non-trivial to implement for an average user and it would be helpful to have it since this thread seems to have concluded that the implementation is trivial.

**Comment by agarwal-ashish** [comment]
One related question is if XLA could fuse reductions with cwise operations to avoid the blowup of memory.

**Comment by cyx96** [comment]
This is the feature I've been waiting for years... Even if it could be easily implemented by users, although I didn't find it very easy, to include this feature in the official library will definitely benefit a lot of average users like me. I'm working on graphs and calculating similarity matrices using TF gave me quite a bit of pain...

**Comment by aewhite** [comment]
I don't know where else to put this so I'll post my finding here. First I did find a pairwise distance function (I think by following linked issues here):

```
def pairwise_dist(A, B):
    na = tf.reduce_sum(tf.square(A), 1)
    nb = tf.reduce_sum(tf.square(B), 1)
    na = tf.reshape(na, [-1, 1])
    nb = tf.reshape(nb, [1, -1])
    D = tf.sqrt(tf.maximum(na - 2 * tf.matmul(A, B, False, True) + nb, 0.0))
    return D
```

If you want something that is maybe more memory friendly then I imitated what scipy does with this setup:

```
def distance(a, b):
    return tf.sqrt(tf.reduce_sum(tf.math.squared_difference(a, b), axis=-1))


def cdist(a, b, metric=None):
    a_i = tf.range(tf.shape(a)[0])
    b_i = tf.range(tf.shape(b)[0])
    indices = tf.concat([tf.expand_dims(x, -1) for x in tf.meshgrid(a_i, b_i, indexing='ij')], axis=-1)
    indices = tf.reshape(indices, [-1, 2])
    distances = tf.map_fn(lambda i: metric(a[i[0]], b[i[1]]), indices, fn_output_signature=tf.float32)
    distances = tf.reshape(distances, [tf.shape(a)[0], tf.shape(b)[0]])
    return distances
```

I highly doubt my solution is optimal but maybe somebody can improve it.

**Comment by E471r** [comment]
All the tools are already there imo:
Same as subtracting two vectors with shapes (n,1) - (1,m) is an (n,m) matrix,
subtracting two matrices X,Y with shapes (m,k,1)-(1,k,n), gives sqeuclidean (m,n) distance matrix by doing:
tf.reduce_sum((X  - tf.einsum('ijk->kji', Y))**2, axis=1)
